File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/realm/models/Shortcut.java
package ch.rmy.android.http_shortcuts.realm.models;

import android.content.Context;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.RequiresApi;

import java.util.List;

import ch.rmy.android.http_shortcuts.R;
import ch.rmy.android.http_shortcuts.utils.ShortcutUIUtils;
import ch.rmy.android.http_shortcuts.utils.Validation;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.Required;

public class Shortcut extends RealmObject implements HasId {

    public static final long TEMPORARY_ID = -1;

    public static final String FIELD_NAME = "name";
    public static final String FIELD_LAUNCHER_SHORTCUT = "launcherShortcut";

    public static final String METHOD_GET = "GET";
    public static final String METHOD_POST = "POST";
    public static final String METHOD_PUT = "PUT";
    public static final String METHOD_DELETE = "DELETE";
    public static final String METHOD_PATCH = "PATCH";
    public static final String METHOD_HEAD = "HEAD";
    public static final String METHOD_OPTIONS = "OPTIONS";
    public static final String METHOD_TRACE = "TRACE";

    public static final String FEEDBACK_NONE = "none";
    public static final String FEEDBACK_TOAST_SIMPLE = "simple_response";
    public static final String FEEDBACK_TOAST_SIMPLE_ERRORS = "simple_response_errors";
    public static final String FEEDBACK_TOAST = "full_response";
    public static final String FEEDBACK_TOAST_ERRORS = "errors_only";
    public static final String FEEDBACK_DIALOG = "dialog";
    public static final String FEEDBACK_ACTIVITY = "activity";

    public static final String RETRY_POLICY_NONE = "none";
    public static final String RETRY_POLICY_WAIT_FOR_INTERNET = "wait_for_internet";

    public static final String[] METHODS = {METHOD_GET, METHOD_POST, METHOD_PUT, METHOD_DELETE, METHOD_PATCH, METHOD_HEAD, METHOD_OPTIONS, METHOD_TRACE};
    public static final String[] FEEDBACK_OPTIONS = {FEEDBACK_NONE, FEEDBACK_TOAST_SIMPLE, FEEDBACK_TOAST_SIMPLE_ERRORS, FEEDBACK_TOAST, FEEDBACK_TOAST_ERRORS, FEEDBACK_DIALOG, FEEDBACK_ACTIVITY};
    public static final int[] TIMEOUT_OPTIONS = {3000, 10000, 30000, 60000};

    public static final String[] RETRY_POLICY_OPTIONS = {RETRY_POLICY_NONE, RETRY_POLICY_WAIT_FOR_INTERNET};

    public static final String AUTHENTICATION_BASIC = "basic";
    public static final String AUTHENTICATION_DIGEST = "digest";

    public static final String[] AUTHENTICATION_OPTIONS = {null, AUTHENTICATION_BASIC, AUTHENTICATION_DIGEST};

    @PrimaryKey
    private long id;
    @Required
    private String name;
    @Required
    private String method = METHOD_GET;
    @Required
    private String url;
    @Required
    private String username;
    @Required
    private String password;
    private String iconName;
    @Required
    private String feedback;
    @Required
    private String description;
    @Required
    private String bodyContent;
    private int timeout;
    @Required
    private String retryPolicy;
    private RealmList<Header> headers;
    private RealmList<Parameter> parameters;
    private boolean acceptAllCertificates;
    private String authentication;
    private boolean launcherShortcut;

    @Override
    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getMethod() {
        return method;
    }

    public void setMethod(String method) {
        this.method = method;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public String getIconName() {
        return iconName;
    }

    public void setIconName(String iconName) {
        this.iconName = iconName;
    }

    public String getFeedback() {
        return feedback;
    }

    public void setFeedback(String feedback) {
        this.feedback = feedback;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getBodyContent() {
        return bodyContent;
    }

    public void setBodyContent(String bodyContent) {
        this.bodyContent = bodyContent;
    }

    public int getTimeout() {
        return timeout;
    }

    public void setTimeout(int timeout) {
        this.timeout = timeout;
    }

    public String getRetryPolicy() {
        return retryPolicy;
    }

    public void setRetryPolicy(String retryPolicy) {
        this.retryPolicy = retryPolicy;
    }

    public RealmList<Header> getHeaders() {
        return headers;
    }

    public void setHeaders(RealmList<Header> headers) {
        this.headers = headers;
    }

    public RealmList<Parameter> getParameters() {
        return parameters;
    }

    public void setParameters(RealmList<Parameter> parameters) {
        this.parameters = parameters;
    }

    public boolean isAcceptAllCertificates() {
        return acceptAllCertificates;
    }

    public void setAcceptAllCertificates(boolean acceptAllCertificates) {
        this.acceptAllCertificates = acceptAllCertificates;
    }

    public String getAuthentication() {
        return authentication;
    }

    public void setAuthentication(String authentication) {
        this.authentication = authentication;
    }

    public boolean isLauncherShortcut() {
        return launcherShortcut;
    }

    public void setLauncherShortcut(boolean launcherShortcut) {
        this.launcherShortcut = launcherShortcut;
    }

    @Override
    public boolean isNew() {
        return id == 0;
    }

    public static Shortcut createNew() {
        Shortcut shortcut = new Shortcut();
        shortcut.setId(0);
        shortcut.setName("");
        shortcut.setDescription("");
        shortcut.setUsername("");
        shortcut.setPassword("");
        shortcut.setBodyContent("");
        shortcut.setMethod(METHOD_GET);
        shortcut.setUrl("http://");
        shortcut.setTimeout(TIMEOUT_OPTIONS[1]);
        shortcut.setFeedback(FEEDBACK_TOAST_SIMPLE);
        shortcut.setRetryPolicy(RETRY_POLICY_NONE);
        shortcut.setParameters(new RealmList<Parameter>());
        shortcut.setHeaders(new RealmList<Header>());
        return shortcut;
    }

    public Shortcut duplicate(String newName) {
        Shortcut duplicate = new Shortcut();
        duplicate.setId(0);
        duplicate.setName(newName);
        duplicate.setBodyContent(getBodyContent());
        duplicate.setDescription(getDescription());
        duplicate.setFeedback(getFeedback());
        duplicate.setIconName(getIconName());
        duplicate.setMethod(getMethod());
        duplicate.setPassword(getPassword());
        duplicate.setRetryPolicy(getRetryPolicy());
        duplicate.setTimeout(getTimeout());
        duplicate.setUrl(getUrl());
        duplicate.setUsername(getUsername());
        duplicate.setAuthentication(getAuthentication());
        duplicate.setLauncherShortcut(isLauncherShortcut());

        duplicate.setParameters(new RealmList<Parameter>());
        for (Parameter parameter : getParameters()) {
            duplicate.getParameters().add(Parameter.createNew(parameter.getKey(), parameter.getValue()));
        }

        duplicate.setHeaders(new RealmList<Header>());
        for (Header header : getHeaders()) {
            duplicate.getHeaders().add(Header.createNew(header.getKey(), header.getValue()));
        }

        return duplicate;
    }

    public Uri getIconURI(Context context) {
        String packageName = context.getPackageName();
        if (iconName == null) {
            return Uri.parse("android.resource://" + packageName + "/" + ShortcutUIUtils.DEFAULT_ICON);
        } else if (iconName.startsWith("android.resource://")) {
            return Uri.parse(iconName);
        } else if (iconName.endsWith(".png")) {
            return Uri.fromFile(context.getFileStreamPath(iconName));
        } else {
            int identifier = context.getResources().getIdentifier(iconName, "drawable", context.getPackageName());
            return Uri.parse("android.resource://" + packageName + "/" + identifier);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public Icon getIcon(Context context) {
        try {
            String packageName = context.getPackageName();
            if (iconName == null) {
                return Icon.createWithResource(packageName, ShortcutUIUtils.DEFAULT_ICON);
            } else if (iconName.startsWith("android.resource://")) {
                List<String> pathSegments = Uri.parse(iconName).getPathSegments();
                return Icon.createWithResource(pathSegments.get(0), Integer.parseInt(pathSegments.get(1)));
            } else if (iconName.endsWith(".png")) {
                return null; // TODO: Generate Icon from Bitmap
            } else {
                int identifier = context.getResources().getIdentifier(iconName, "drawable", context.getPackageName());
                return Icon.createWithResource(packageName, identifier);
            }
        } catch (Exception e) {
            return null;
        }
    }

    public String getSafeName(Context context) {
        if (Validation.isEmpty(name)) {
            return context.getString(R.string.shortcut_safe_name);
        }
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Shortcut shortcut = (Shortcut) o;

        if (getTimeout() != shortcut.getTimeout()) return false;
        if (!getName().equals(shortcut.getName())) return false;
        if (!getMethod().equals(shortcut.getMethod())) return false;
        if (!getUrl().equals(shortcut.getUrl())) return false;
        if (!getUsername().equals(shortcut.getUsername())) return false;
        if (!getPassword().equals(shortcut.getPassword())) return false;
        if (getIconName() != null ? !getIconName().equals(shortcut.getIconName()) : shortcut.getIconName() != null)
            return false;
        if (getAuthentication() != null ? !getAuthentication().equals(shortcut.getAuthentication()) : shortcut.getAuthentication() != null)
            return false;
        if (!getFeedback().equals(shortcut.getFeedback())) return false;
        if (!getDescription().equals(shortcut.getDescription())) return false;
        if (!getBodyContent().equals(shortcut.getBodyContent())) return false;
        if (!getRetryPolicy().equals(shortcut.getRetryPolicy())) return false;
        if (isLauncherShortcut() != shortcut.isLauncherShortcut()) return false;
        if (!getHeaders().equals(shortcut.getHeaders())) return false;
        return getParameters().equals(shortcut.getParameters());
    }

    @Override
    public int hashCode() {
        int result = (int) (getId() ^ (getId() >>> 32));
        result = 31 * result + getName().hashCode();
        result = 31 * result + getMethod().hashCode();
        result = 31 * result + getUrl().hashCode();
        result = 31 * result + getUsername().hashCode();
        result = 31 * result + getPassword().hashCode();
        result = 31 * result + (getIconName() != null ? getIconName().hashCode() : 0);
        result = 31 * result + (getAuthentication() != null ? getAuthentication().hashCode() : 0);
        result = 31 * result + getFeedback().hashCode();
        result = 31 * result + getDescription().hashCode();
        result = 31 * result + getBodyContent().hashCode();
        result = 31 * result + getTimeout();
        result = 31 * result + getRetryPolicy().hashCode();
        result = 31 * result + getHeaders().hashCode();
        result = 31 * result + getParameters().hashCode();
        result = 31 * result + (isLauncherShortcut() ? 42 : 0);
        return result;
    }

    public boolean allowsBody() {
        return METHOD_POST.equals(getMethod())
                || METHOD_PUT.equals(getMethod())
                || METHOD_DELETE.equals(getMethod())
                || METHOD_PATCH.equals(getMethod())
                || METHOD_OPTIONS.equals(getMethod());
    }

    public boolean feedbackUsesUI() {
        return FEEDBACK_DIALOG.equals(getFeedback())
                || FEEDBACK_ACTIVITY.equals(getFeedback());
    }

    public boolean isFeedbackErrorsOnly() {
        return FEEDBACK_TOAST_ERRORS.equals(getFeedback())
                || FEEDBACK_TOAST_SIMPLE_ERRORS.equals(getFeedback());
    }

    public boolean isRetryAllowed() {
        return !FEEDBACK_ACTIVITY.equals(getFeedback())
                && !FEEDBACK_DIALOG.equals(getFeedback());
    }

    public boolean usesAuthentication() {
        return usesBasicAuthentication()
                || usesDigestAuthentication();
    }

    public boolean usesBasicAuthentication() {
        return AUTHENTICATION_BASIC.equals(getAuthentication());
    }

    public boolean usesDigestAuthentication() {
        return AUTHENTICATION_DIGEST.equals(getAuthentication());
    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/realm/models/Shortcut.kt
package ch.rmy.android.http_shortcuts.realm.models

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import android.support.annotation.RequiresApi
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.utils.ShortcutUIUtils
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import io.realm.annotations.Required

open class Shortcut : RealmObject(), HasId {

    @PrimaryKey
    override var id: Long = 0
    @Required
    var name: String = ""
    @Required
    var method = METHOD_GET
    @Required
    var url: String = ""
    @Required
    var username: String = ""
    @Required
    var password: String = ""
    var iconName: String? = null
    @Required
    var feedback: String = ""
    @Required
    var description: String = ""
    @Required
    var bodyContent: String = ""
    var timeout: Int = 0
    @Required
    var retryPolicy: String = ""
    var headers: RealmList<Header> = RealmList()
    var parameters: RealmList<Parameter> = RealmList()
    var acceptAllCertificates: Boolean = false
    var authentication: String? = AUTHENTICATION_NONE
    var launcherShortcut: Boolean = false
    var delay: Int = 0
    @Required
    var requestBodyType: String = REQUEST_BODY_TYPE_CUSTOM_TEXT
    @Required
    var contentType: String = ""

    var beforeActions: String? = null
    var successActions: String? = null
    var failureActions: String? = null

    override val isNew: Boolean
        get() = id == 0L

    fun duplicate(newName: String): Shortcut {
        val duplicate = Shortcut()
        duplicate.id = 0
        duplicate.name = newName
        duplicate.bodyContent = bodyContent
        duplicate.description = description
        duplicate.feedback = feedback
        duplicate.iconName = iconName
        duplicate.method = method
        duplicate.password = password
        duplicate.retryPolicy = retryPolicy
        duplicate.timeout = timeout
        duplicate.url = url
        duplicate.username = username
        duplicate.authentication = authentication
        duplicate.launcherShortcut = launcherShortcut
        duplicate.acceptAllCertificates = acceptAllCertificates
        duplicate.delay = delay
        duplicate.requestBodyType = requestBodyType
        duplicate.contentType = contentType
        duplicate.beforeActions = beforeActions
        duplicate.successActions = successActions
        duplicate.failureActions = failureActions

        duplicate.parameters = RealmList()
        for (parameter in parameters) {
            duplicate.parameters.add(Parameter.createNew(parameter.key, parameter.value))
        }

        duplicate.headers = RealmList()
        for (header in headers) {
            duplicate.headers.add(Header.createNew(header.key, header.value))
        }

        return duplicate
    }

    fun getIconURI(context: Context): Uri {
        val packageName = context.packageName
        return when {
            iconName == null -> Uri.parse("android.resource://" + packageName + "/" + ShortcutUIUtils.DEFAULT_ICON)
            iconName!!.startsWith("android.resource://") -> Uri.parse(iconName)
            iconName!!.endsWith(".png") -> Uri.fromFile(context.getFileStreamPath(iconName))
            else -> {
                val identifier = context.resources.getIdentifier(iconName, "drawable", context.packageName)
                Uri.parse("android.resource://$packageName/$identifier")
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    fun getIcon(context: Context): Icon? = try {
        val packageName = context.packageName
        when {
            iconName == null -> Icon.createWithResource(packageName, ShortcutUIUtils.DEFAULT_ICON)
            iconName!!.startsWith("android.resource://") -> {
                val pathSegments = Uri.parse(iconName).pathSegments
                Icon.createWithResource(pathSegments[0], Integer.parseInt(pathSegments[1]))
            }
            iconName!!.endsWith(".png") -> {
                val options = BitmapFactory.Options()
                options.inPreferredConfig = Bitmap.Config.ARGB_8888
                val bitmap = BitmapFactory.decodeFile(context.getFileStreamPath(iconName).absolutePath, options)
                Icon.createWithBitmap(bitmap)
            }
            else -> {
                val identifier = context.resources.getIdentifier(iconName, "drawable", context.packageName)
                Icon.createWithResource(packageName, identifier)
            }
        }
    } catch (e: Exception) {
        null
    }

    fun getSafeName(context: Context): String = if (name.isBlank()) {
        context.getString(R.string.shortcut_safe_name)
    } else {
        name
    }

    fun allowsBody(): Boolean {
        return METHOD_POST == method
                || METHOD_PUT == method
                || METHOD_DELETE == method
                || METHOD_PATCH == method
                || METHOD_OPTIONS == method
    }

    fun feedbackUsesUI() = feedback == FEEDBACK_DIALOG || feedback == FEEDBACK_ACTIVITY

    fun isFeedbackErrorsOnly() =
            feedback == FEEDBACK_TOAST_ERRORS || feedback == FEEDBACK_TOAST_SIMPLE_ERRORS

    fun isRetryAllowed() = feedback != FEEDBACK_ACTIVITY && feedback != FEEDBACK_DIALOG

    fun usesAuthentication() = usesBasicAuthentication() || usesDigestAuthentication()

    fun usesBasicAuthentication() = authentication == AUTHENTICATION_BASIC

    fun usesDigestAuthentication() = authentication == AUTHENTICATION_DIGEST

    fun usesRequestParameters() = allowsBody() && (requestBodyType == REQUEST_BODY_TYPE_FORM_DATA || requestBodyType == REQUEST_BODY_TYPE_X_WWW_FORM_URLENCODE)

    fun usesCustomBody() = allowsBody() && requestBodyType == REQUEST_BODY_TYPE_CUSTOM_TEXT

    fun isSameAs(other: Shortcut): Boolean {
        if (other.name != name ||
                other.bodyContent != bodyContent ||
                other.description != description ||
                other.feedback != feedback ||
                other.iconName != iconName ||
                other.method != method ||
                other.password != password ||
                other.retryPolicy != retryPolicy ||
                other.timeout != timeout ||
                other.url != url ||
                other.username != username ||
                other.authentication != authentication ||
                other.launcherShortcut != launcherShortcut ||
                other.acceptAllCertificates != acceptAllCertificates ||
                other.delay != delay ||
                other.parameters.size != parameters.size ||
                other.headers.size != headers.size ||
                other.requestBodyType != requestBodyType ||
                other.contentType != contentType ||
                other.beforeActions != beforeActions ||
                other.successActions != successActions ||
                other.failureActions != failureActions
        ) {
            return false
        }
        if (other.parameters.indices.any { !parameters[it]!!.isSameAs(other.parameters[it]!!) }) {
            return false
        }
        if (other.headers.indices.any { !headers[it]!!.isSameAs(other.headers[it]!!) }) {
            return false
        }
        return true
    }

    companion object {

        const val TEMPORARY_ID: Long = -1

        const val FIELD_NAME = "name"

        const val METHOD_GET = "GET"
        const val METHOD_POST = "POST"
        const val METHOD_PUT = "PUT"
        const val METHOD_DELETE = "DELETE"
        const val METHOD_PATCH = "PATCH"
        const val METHOD_HEAD = "HEAD"
        const val METHOD_OPTIONS = "OPTIONS"
        const val METHOD_TRACE = "TRACE"

        const val FEEDBACK_NONE = "none"
        const val FEEDBACK_TOAST_SIMPLE = "simple_response"
        const val FEEDBACK_TOAST_SIMPLE_ERRORS = "simple_response_errors"
        const val FEEDBACK_TOAST = "full_response"
        const val FEEDBACK_TOAST_ERRORS = "errors_only"
        const val FEEDBACK_DIALOG = "dialog"
        const val FEEDBACK_ACTIVITY = "activity"

        const val RETRY_POLICY_NONE = "none"
        const val RETRY_POLICY_WAIT_FOR_INTERNET = "wait_for_internet"

        const val REQUEST_BODY_TYPE_FORM_DATA = "form_data"
        const val REQUEST_BODY_TYPE_X_WWW_FORM_URLENCODE = "x_www_form_urlencode"
        const val REQUEST_BODY_TYPE_CUSTOM_TEXT = "custom_text"

        val METHODS = arrayOf(METHOD_GET, METHOD_POST, METHOD_PUT, METHOD_DELETE, METHOD_PATCH, METHOD_HEAD, METHOD_OPTIONS, METHOD_TRACE)
        val FEEDBACK_OPTIONS = arrayOf(FEEDBACK_NONE, FEEDBACK_TOAST_SIMPLE, FEEDBACK_TOAST_SIMPLE_ERRORS, FEEDBACK_TOAST, FEEDBACK_TOAST_ERRORS, FEEDBACK_DIALOG, FEEDBACK_ACTIVITY)
        val TIMEOUT_OPTIONS = intArrayOf(3000, 10000, 30000, 60000)
        val DELAY_OPTIONS = intArrayOf(0, 5000, 10000, 30000, 60000, 120000, 300000, 600000)

        val RETRY_POLICY_OPTIONS = arrayOf(RETRY_POLICY_NONE, RETRY_POLICY_WAIT_FOR_INTERNET)

        const val AUTHENTICATION_NONE = "none"
        const val AUTHENTICATION_BASIC = "basic"
        const val AUTHENTICATION_DIGEST = "digest"

        val AUTHENTICATION_OPTIONS = arrayOf(AUTHENTICATION_NONE, AUTHENTICATION_BASIC, AUTHENTICATION_DIGEST)

        val REQUEST_BODY_TYPE_OPTIONS = arrayOf(REQUEST_BODY_TYPE_FORM_DATA, REQUEST_BODY_TYPE_X_WWW_FORM_URLENCODE, REQUEST_BODY_TYPE_CUSTOM_TEXT)

        val CONTENT_TYPE_SUGGESTIONS = arrayOf(
                "application/javascript",
                "application/json",
                "application/octet-stream",
                "application/xml",
                "text/css",
                "text/csv",
                "text/plain",
                "text/html",
                "text/xml"
        )

        const val DEFAULT_CONTENT_TYPE = "text/plain"

        fun createNew() = Shortcut().apply {
            id = 0
            name = ""
            description = ""
            username = ""
            password = ""
            bodyContent = ""
            method = METHOD_GET
            url = "http://"
            timeout = TIMEOUT_OPTIONS[1]
            feedback = FEEDBACK_TOAST_SIMPLE
            retryPolicy = RETRY_POLICY_NONE
            authentication = AUTHENTICATION_NONE
            delay = 0
            parameters = RealmList()
            headers = RealmList()
            requestBodyType = REQUEST_BODY_TYPE_X_WWW_FORM_URLENCODE
            contentType = DEFAULT_CONTENT_TYPE
            beforeActions = "[]"
            successActions = "[]"
            failureActions = "[]"
        }
    }

}
